src/openassetio-python/trait/BlobTrait.cpp

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#include <optional>
+#include <type_traits>
+
+#include <pybind11/stl.h>
+
+// Public.
+#include <openassetio/export.h>
+#include <openassetio/trait/BlobTrait.hpp>
+// Private.
+#include <trait/keys.hpp>
```

Comment:
[question] Here we're `#include`ing a header that is not part of the public API, i.e. not in the `include` directory.  

To make this work, the build target's include directories must overlap, i.e. both the `openassetio-core` directory and the `openassetio-core/include` directory are on the search path.  

I think the directory structure should be rejigged to remove the overlap, but I can't decide a good split.

Currently we have
```
openassetio-core
├── include
│   └── openassetio
│       ├── managerAPI
│       ├── specification
│       └── trait
├── managerAPI
├── specification
└── trait
```

We could e.g. restructure to something like
```
openassetio-core
├── include
│   └── openassetio
│       ├── managerAPI
│       ├── specification
│       └── trait
└── src
    └── private
        ├── managerAPI
        ├── specification
        └── trait
```
which would allow `#include <private/trait/keys.hpp>` (assuming `src` was on the include search path).  

I don't particularly like this option, but it's surprisingly hard to come up with a a scheme that doesn't overlap, isn't overly verbose, gives clear at-a-glance `#include`s, whilst (ideally) dog-fooding our use of public includes.

We can look at other projects to get some inspiration.

Reply:
How come its not part of the public API? If it's needed for a custom trait, then it needs to be public - as traits have to be runtime extensible?

Reply:
I think I just answered that Q when looking into the comment above (sorry was visiting these in reverse order), either way, we should probably have a huddle on src layout?